antipackage.py

```diff
@@ -1,130 +1,401 @@
+        pathlist = self.index[1:]
+        reponame = '.'.join(pathlist)
+        for key in self.data:
+            if key in ('branch','tag'):
+                repo = GitHubRepo(pathlist, key, self.data[key])
+                if repo.sha != self.data['commit']['sha']:
+                    print('Updating repo: %s' % reponame)
+                    self.data['commit']['sha'] = repo.sha
+                    return True
+        print('Using existing version of: %s' % reponame)
+        return False
+
+    def install(self):
+        """Download and install new repository at self.path"""
+        url = self.zip()
+        print("Downloading from: %s" % url)
+        sys.stdout.flush()
+        temp_file, response = urlretrieve(url, reporthook=display_download)
+        zf = zipfile.ZipFile(temp_file)
+        fname = zf.filename
+        with TemporaryDirectory() as td:
+            zf.extractall(td)
+            contents = os.listdir(td)
+            for name in contents:
+                if name.startswith(self.username):
+                    tmp_repo = name
+            if os.path.exists(self.path):
+                shutil.rmtree(self.path)
+            os.rename(td+'/'+tmp_repo, self.path)
+        self._install_init()
+
+    def zip(self):
+        """Returns url to the reopsitory's .zip file"""
+        path = ['repos',self.username, self.repo,'zipball']
+        if self.data:
+            src = self.data['commit']['sha']
+        else:
+            pin('/'.join(self.index[1:]))
+            src = 'master'
+        path.append(src)
+        return 'https://api.github.com/'+'/'.join(path)
+
+class GitHubRepo(object):
+
+    host = {'top': 'https://api.github.com/repos/{1}/{2}/',
+            'branch': 'branches/{0}',
+            'sha': 'commits/{0}',
+            'tag': 'tags'}
+
+    def __init__(self, pathlist, name='branch', value='master'):
+        """Stores key GitHub repository information
+
+        Parameters
+        ----------
+        pathlist : list
+            Of the form `[github, username, repository]`
+            and indicates the repository that should be
+            accessed
+        name : str
+            Should be either 'sha', 'tag' or 'branch' and
+            indicates the version access method for the repo
+        value : str
+            specific to `name` and indicated the specific
+            repository version being referenced"""
+        self.name = name
+        self.value = value
+        self.url = self._url(pathlist)
+        self._git_init()
+
+    def _url(self, pathlist):
+        """Format the GitHub url"""
+        url = self.host['top'].format(*pathlist)
+        url += self.host[self.name].format(self.value)
+        return url
+
+    def _git_init(self):
+        """Initialize data from GitHub"""
+        temp_file = self.fetch()
+        with open(temp_file, 'r') as f:
+            data = json.load(f)
+            self.sha = self._sha(data)
+
+    def fetch(self):
```

Comment:
Let's move this up right below the `__init__` method to keep the public methods together.